holmes/plugins/toolsets/kubernetes_logs.py

```diff
@@ -0,0 +1,281 @@
+
+from holmes.core.tools import (
+    StaticPrerequisite,
+    StructuredToolResult,
+    ToolResultStatus,
+    ToolsetTag,
+)
+from holmes.plugins.toolsets.logging_api import (
+    BaseLoggingToolset,
+    FetchLogsParams,
+    LoggingConfig,
+    LoggingTool,
+)
+from holmes.plugins.toolsets.utils import process_timestamps_to_int, to_unix
+
+
+class Pod(BaseModel):
+    containers: list[str]
+
+
+class KubernetesLogsToolset(BaseLoggingToolset):
+    """Implementation of the unified logging API for Kubernetes logs using the official Python client"""
+
+    def __init__(self):
+        prerequisite = StaticPrerequisite(enabled=False, disabled_reason="Initializing")
+        super().__init__(
+            name="kubernetes/logs",
+            description="Read Kubernetes pod logs using a unified API",
+            docs_url="https://docs.robusta.dev/master/configuration/holmesgpt/toolsets/kubernetes.html#logs",
+            icon_url="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRPKA-U9m5BxYQDF1O7atMfj9EMMXEoGu4t0Q&s",
+            prerequisites=[prerequisite],
+            tools=[
+                LoggingTool(self),
+            ],
+            tags=[ToolsetTag.CORE],
+        )
+        self._api_client = None
+        self._core_v1_api = None
+        self._initialize_client()
+        enabled, disabled_reason = self.health_check()
+        prerequisite.enabled = enabled
+        prerequisite.disabled_reason = disabled_reason
+
+    def health_check(self) -> Tuple[bool, str]:
+        if self._api_client is None:
+            return False, "Kubernetes client not initialized"
+        try:
+            # Try to load the kubeconfig and access the API
+            self._api_client.call_api(
+                "/version",
+                "GET",
+                auth_settings=["BearerToken"],
+                response_type="object",
+                _return_http_data_only=True,
+            )
+            return True, ""
+        except Exception as e:
+            return False, f"Kubernetes client initialization error: {str(e)}"
+
+    def get_example_config(self):
+        return LoggingConfig().model_dump()
+
+    def _initialize_client(self):
+        try:
+            try:
+                config.load_incluster_config()
+            except config.ConfigException:
+                logging.debug(
+                    f"_initialize_client for {self.name} toolset falling back to loading kube_config"
+                )
+                config.load_kube_config()
+
+            self._api_client = client.ApiClient()
+            self._core_v1_api = client.CoreV1Api(self._api_client)
+        except Exception:
+            logging.error("Failed to initialize Kubernetes client", exc_info=True)
+
+    def fetch_logs(self, params: FetchLogsParams) -> StructuredToolResult:
+        try:
+            pod = self._find_pod(params.namespace, params.pod_name)
+            all_logs = []
+            if not pod or not pod.containers:
```

Comment:
Why we want to fetch logs if 'not pod'? e.g. if we couldn't find the pod how we can fetch it logs?

Reply:
A terminated pod may return nothing when `_find_pod` is called.

Reply:
yes, sometimes